azerothcore/azerothcore-wotlk
# feat(Core/Hooks): Add some script hooks for custom module.

## Changes Proposed:

- Implement a hook called after player move worldport. (teleport between two maps)
- Implement a hook called after player move teleport. (teleport within map)
- Implement a hook called from beginning of creature Update tick.
- Implement a hook called after creature finished cast a spell.

## Tests Performed:

- Compiled with test module 'mod-npc-bots' at Windows 10, ubuntu and macOS(Big Sur).
- Summon a bot in game, then attack a creature. make sure hook 'OnSpellGo' called correctly.
- Player teleport to another map, make sure hook 'OnPlayerMoveWorldport' called correctly.
- Player teleport to another area in the same map, make sure hook 'OnPlayerMoveTeleport' called correctly.

## How to Test the Changes:


1. Try hooking OnPlayerMoveWorldport(Player* player) , make sure everything is working.
2. Try hooking OnPlayerMoveTeleport(Player* player) , make sure everything is working.
3. Try hooking OnBeforeCreatureUpdate(Creature*, uint32) , make sure everything is working.
4. Try hooking OnSpellGo(Unit const* caster, Spell const* spell, bool ok) , make sure everything is working.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
